packages/core/src/modules/credentials/v2/CredentialMessageBuilder.ts

```diff
@@ -0,0 +1,273 @@
+import type { CredentialRecordProps } from '../repository/CredentialRecord'
+import type { V2CredentialPreview } from './V2CredentialPreview'
+import type { CredentialFormatService } from './formats/CredentialFormatService'
+import type {
+  AcceptProposalOptions,
+  ProposeCredentialOptions,
+  RequestCredentialOptions,
+  V2CredOfferFormat,
+  V2CredRequestFormat,
+} from './interfaces'
+import type { V2ProposeCredentialMessageOptions } from './messages/V2ProposeCredentialMessage'
+import type { V2RequestCredentialMessageOptions } from './messages/V2RequestCredentialMessage'
+import type { AgentMessage } from 'packages/core/src/agent/AgentMessage'
+
+import { CredentialState } from '..'
+import { AriesFrameworkError } from '../../../error'
+import { isLinkedAttachment } from '../../../utils/attachment'
+import { CredentialRecord } from '../repository/CredentialRecord'
+
+import { V2OfferCredentialMessage } from './messages/V2OfferCredentialMessage'
+import { V2ProposeCredentialMessage } from './messages/V2ProposeCredentialMessage'
+import { V2RequestCredentialMessage } from './messages/V2RequestCredentialMessage'
+
+export interface CredentialProtocolMsgReturnType<MessageType extends AgentMessage> {
+  message: MessageType
+  credentialRecord: CredentialRecord
+}
+
+export class CredentialMessageBuilder {
```

Comment:
The credential service feels quite light at the moment. I'm in favour of adding extra abstraction layers (such as the `CredentialMessageBuilder`), but I'm not sure if this is the right implementation of it. It is not clear to me exactly what should go into the CredentielMessageBuilder vs the V2CredentialService, and some methods have nothing to do with building messages (e.g. `getCredentialDefinitionId`). 

The methods that call this builder have a very small function body because most of the logic happens here, so my preference would be took keep most of the logic in the `V2CredentialService` for now. I'm open to other suggestions

Reply:
I have been grappling with this concept as well, ie functionality in the CredentialsService vs the MessageBuilder.

I think the message builder has grown more than I wanted. I will make these changes later.